proxy/src/main/java/com/wavefront/agent/TenantInfo.java

```diff
@@ -0,0 +1,196 @@
+package com.wavefront.agent;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.net.HttpHeaders;
+import jakarta.ws.rs.client.Entity;
+import jakarta.ws.rs.core.Form;
+import jakarta.ws.rs.core.Response;
+import org.glassfish.jersey.client.JerseyClientBuilder;
+
+import javax.annotation.Nonnull;
+import javax.ws.rs.core.MediaType;
+import java.util.Base64;
+import java.util.concurrent.TimeUnit;
+
+import static com.wavefront.agent.ProxyConfig.PROXY_AUTH_METHOD.CSP_API_TOKEN;
+import static jakarta.ws.rs.core.Response.Status.Family.SUCCESSFUL;
+import static java.lang.String.format;
+
+/**
+ * The class for keeping tenant required information token, server.
+ *
+ * @author Norayr Chaparyan (nchaparyan@vmware.com).
+ */
+public class TenantInfo {
+    private static final String GET_CSP_ACCESS_TOKEN_ERROR_MESSAGE = "Failed to get access token from CSP.";
+    private static final String CSP_AUTH_TOKEN = "csp-auth-token";
+    // The CSP console url.
+    private static final String cspBaseUrl = "https://console-stg.cloud.vmware.com";
+
+    // This can be a Wavefront API token or a CSP API token.
+    private String tenantToken;
+    private final String tenantServer;
+    private final LoadingCache<String, TokenExchangeResponseDTO> tokenCache;
+
+    TenantInfo(@Nonnull final String tenantToken,
+               @Nonnull final String tenantServer,
+               ProxyConfig.PROXY_AUTH_METHOD proxyAuthMethod) {
+        this.tenantToken = tenantToken;
+        this.tenantServer = tenantServer;
+        if (proxyAuthMethod == CSP_API_TOKEN) {
+            tokenCache = Caffeine.newBuilder().
```

Comment:
if this will have only one item, why use a Caffeine cache, looks overkill.

Reply:
> if this will have only one item, why use a Caffeine cache, looks overkill.

We use Caffeine cache for regularly refreshing token if it comes from CSP based on token TTL.

Reply:
fyi, you might review `Suppliers.memoizeWithExpiration` which uses double checking locking with a TTL check. Their api has the TTL being static, but conceptually would not need to be in a custom variant.

Reply:
Done, thanks